src/artemis/zocalo_interaction.py

```diff
@@ -49,3 +62,47 @@ def run_end(data_collection_id: int):
             "ispyb_dcid": data_collection_id,
         }
     )
+
+
+def wait_for_result(data_collection_id: int, timeout: int = TIMEOUT) -> Point3D:
+    """Block until a result is received from Zocalo.
+    Args:
+        data_collection_id (int): The ID of the data collection representing the gridscan in ISpyB
+        timeout (float): The time in seconds to wait for the result to be received.
+    Returns:
+        Point in grid co-ordinates that is the centre point to move to
+    """
+    transport = _get_zocalo_connection()
+    result_received = None
+
+    def receive_result(
+        rw: workflows.recipe.RecipeWrapper, header: dict, message: dict
+    ) -> None:
+        print(f"Received {message}")
+        recipe_parameters = rw.recipe_step["parameters"]
+        print(f"Recipe step parameters: {recipe_parameters}")
+        transport.ack(header)
+        if recipe_parameters["dcid"] == str(data_collection_id):
+            nonlocal result_received
+            result_received = Point3D(*message["max_voxel"])
+        else:
+            print(
+                f"Warning: results for {recipe_parameters['dcid']} received but expected {data_collection_id}"
+            )
+
+    workflows.recipe.wrap_subscribe(
+        transport,
+        "xrc.i03",
+        receive_result,
+        acknowledgement=True,
+        allow_non_recipe_messages=False,
+    )
+
+    try:
+        for _ in range(timeout):
+            sleep(1)
+            if result_received:
+                return result_received
+        raise TimeoutError(f"Result not received in {timeout} seconds")
+    finally:
+        transport.disconnect()
```

Comment:
I don't think this use of `nonlocal` to communicate the result back to the main thread will work as the subscription callback will be called in another thread. Something along the lines of the following, using `queue.Queue` to communicate the result works in my test consumer:

```suggestion
    results_queue: queue.Queue = queue.Queue()

    def receive_result(
        rw: workflows.recipe.RecipeWrapper, header: dict, message: dict
    ) -> None:
        print(f"Received {message}")
        recipe_parameters = rw.recipe_step["parameters"]
        print(f"Recipe step parameters: {recipe_parameters}")
        transport.ack(header)
        if recipe_parameters["dcid"] == str(data_collection_id):
            results_queue.put(max_voxel)
        else:
            print(
                f"Warning: results for {recipe_parameters['dcid']} received but expected {data_collection_id}"
            )

    workflows.recipe.wrap_subscribe(
        transport,
        "xrc.i03",
        receive_result,
        acknowledgement=True,
        allow_non_recipe_messages=False,
    )

    max_voxel = None
    wait = 0.1  # seconds
    start_time = time.time()
    while True:
        try:
            max_voxel = Point3D(*results_queue.get(True, wait))
            print(f"{max_voxel=}")
            break
        except queue.Empty:
            if time.time() - start_time >= timeout:
                print(f"Timed out after {timeout} seconds")
                break
            time.sleep(wait)
            continue
    transport.disconnect()
    if max_voxel is None:
        raise TimeoutError(f"Result not received in {timeout} seconds")
    return max_voxel
```

Reply:
Great, thanks. I've added a `queue, I think we can rely on the `queue` itself to do the timeout, rather than having to do it ourselves?

Reply:
Yes, I think this should be fine if you're just expecting to receive a single result.